lib/pos_blockchain.go

```diff
@@ -76,17 +77,32 @@ func (bc *Blockchain) processBlockPoS(block *MsgDeSoBlock, currentView uint64, v
 		return false, false, nil, errors.Wrap(err, "processBlockPoS: Problem applying new tip: ")
 	}
 
-	// If the incoming block is not applied as the new tip, we return early.
-	if !appliedNewTip {
-		return false, false, nil, nil
+	// 5. Commit grandparent if possible. Only need to do this if we applied a new tip.
+	if appliedNewTip {
+		if err = bc.runCommitRuleOnBestChain(); err != nil {
+			return false, false, nil, errors.Wrap(err, "processBlockPoS: error running commit rule: ")
+		}
 	}
 
-	// 5. Commit grandparent if possible.
-	if err = bc.runCommitRuleOnBestChain(); err != nil {
-		return false, false, nil, errors.Wrap(err, "processBlockPoS: error committing grandparents: ")
+	// Now that we've processed this block, we check for any blocks that were previously
+	// stored as orphans, which are children of this block. We can  process them now.
+	blockNodesAtNextHeight := bc.blockIndexByHeight[uint64(blockNode.Height)+1]
+	for _, blockNodeAtNextHeight := range blockNodesAtNextHeight {
+		if blockNodeAtNextHeight.Header.PrevBlockHash.IsEqual(blockNode.Hash) && blockNodeAtNextHeight.IsStored() && !blockNodeAtNextHeight.IsValidated() {
+			var orphanBlock *MsgDeSoBlock
+			orphanBlock, err = GetBlock(blockNodeAtNextHeight.Hash, bc.db, bc.snapshot)
+			if err != nil {
+				glog.Errorf("processBlockPoS: Problem getting orphan block %v", blockNodeAtNextHeight.Hash)
+				continue
+			}
+			if _, _, _, err = bc.processBlockPoS(orphanBlock, currentView, verifySignatures); err != nil {
```

Comment:
I think you still need to capture whether one of the recursive processBlockPoS calls applied the new tip, no? If one of the recursive calls applied a new tip, then it means the current execution of processBlockPoS resulted in a new tip.

And we need to reflect that via the `appliedNewTip` return value at the bottom of this function. I'd pretty define the `appliedNewTip` return value to mean: "we processed this block and as a result, the tip for the best chain has changed to some new tip"

Reply:
oooh good call. will update.